core/dbt/include/global_project/macros/utils/listagg.sql

```diff
@@ -0,0 +1,30 @@
+{% macro listagg(measure, delimiter_text="','", order_by_clause=none, limit_num=none) -%}
```

Comment:
In my work on https://github.com/dbt-labs/dbt-spark/pull/359, I've realized that `order_by_clause` is really not supported on the SparkSQL analogue for `listagg`. I don't think that's a reason to exclude it from the standard / default implementation, since so many other SQL dialects support it so handily, but:
- We _might_ be able to support it, via some array sorting, if it were expressed as `order_by=<column_name>`, instead of expecting a SQL `order by column_name` expression
- I'm curious what you think about the approach I took [here](https://github.com/dbt-labs/dbt-spark/pull/359/files#diff-767481c1a1c644d78d4df7e35bb740cbe0b8fb65ce3eeadeb71f93300058037eR3-R5) — basically, raise a warning if the argument is passed, then ignore it

Reply:
The approach you took with ignoring the `order_by_clause` parameter and raising a warning is completely reasonable.

Definitely worth a look if we can replace a user-supplied value like `order by column_name expression` with `order_by=<column_name>` instead. I'll take a peek and report back.

Reply:
Overall, I think we should defer this work to a separate PR (if we do in fact take it on).

Before taking it on, here are the main things for us to answer:
1. Do all databases operate correctly with an expression like `order_by_clause="order by column_name"`?
2. Would it be a breaking change? If so, is there any path forward that is non-breaking?

## Findings for question 1
Summary: Yes, it looks like they all support `order_by_clause="order by column_name"`

Details:
- [Here](https://github.com/dbt-labs/dbt-utils/blob/ac072a3c4b78d43a1c013e7de8b8fa6e290b544e/macros/cross_db_utils/listagg.sql) are current implementations across databases.
    - Looks like our supported adapters each needs a different implementation, possibly because it was a relative late-comer first appearing in SQL:2016. (Fascinating discussion on `listagg` [here](https://modern-sql.com/feature/listagg))
- The pytests [here](https://github.com/dbt-labs/dbt-core/blob/1001e42bdd686c85b4191d943d5089d9af333d79/tests/adapter/dbt/tests/adapter/utils/fixture_listagg.py) include expressions like `"order by order_col"` (exactly as you expected)
- The pytests use positional arguments exclusively and don't currently cover named keyword arguments.
    - 🟢 Covering the positional arguments is most crucial since messing with order of those can affect consumers relying on a particular order.
    - 🟡 We might want to add keyword testing for good measure though.
    
## Findings for question 2
It _would_ be a breaking change if we drop `order_by_clause="order by column_name"` and replace it with something like `order_by="column_name"`.

Options:
1. Accept the breaking change
2. Reject the breaking change
3. Introduce the new parameter and deprecate the old parameter in a non-breaking fashion.

The last option would look something like this:
- New signature:`{% macro listagg(measure, delimiter_text="','", order_by_clause=none, limit_num=none, order_by=none) -%}`
    - Doing it this way prevents unexpected errors for any consumers relying upon positional arguments
- Raise an error (or warn) if both `order_by_clause` and `order_by` are provided
- When only `order_by_clause` is provided:
    - This is the "base case", and all the current code is relevant
    - Provide a deprecation warning that this parameter will be removed in a future release.
- When only `order_by_clause` is provided:
    - For non-spark adapters, convert `order_by` into the relevant `order_by_clause` value. Proceed to the code in the base case above.
    - For spark, use `order_by` according to your plans 😈 
    
When doing the final deprecation in the future, we'd probably want to leave `order_by_clause=none` as a non-functional vestigial parameter to preserve the positions of each keyword parameter.

Reply:
@dbeatty10 Thanks for going deep on this—much deeper than I did!

I did some more poking, and the amount of `struct` + `array_sort` nonsense that would be required even just to get `order_by=column_name` working is nontrivial. More effort than it's worth until we have a proven need. It's an optional argument, and feels like a nice-to-have.

I'm quite happy to leave this in place, and document / warn that `order_by_clause` is not supported on SparkSQL. It does seem to be a (recent) standard, and most other databases support it in exactly this form (`order_by_clause`). Perhaps we could convince the fine folks developing Apache Spark to add support for it, someday? :)